src/pyhf/schema/upgrader.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import annotations
+
+from pyhf.typing import Workspace, PatchSet, SchemaVersion, UpgradeProtocol
+import copy
+
+
+class Upgrade_1_0_1:
+    """
+    Used for testing functionality of upgrade.
+    """
+
+    version: SchemaVersion = '1.0.1'
```

Comment:
```suggestion
class Upgrade(version: SchemaVersion = "1.0.1"):
    """
    Used for testing functionality of upgrade.
    """
```

Maybe I'm missing something about what the purpose of `upgrader` is, but is there a reason why we need to make this class schema `v1.0.1` hardcoded?

Reply:
We need an `Upgrade_X_Y_Z` class for every single version we plan to support upgrading to. When we add a new version of the spec, we have to have a new upgrade class that knows how to upgrade from an old version to the version the class represents.

Reply:
@kratsg Ah okay, so the real reason is that we can't deterministically know what version we're upgrading from in advance? 

Do we have a plan for how to handle upgrading multiple versions in one go? e.g. Let's say we have `v1.0.0`, `v1.1.0`, `v1.2.0, and `v2.0.0`. If at this point someone comes along with a `v1.0.0` schema and wants to move from `v1.0.0` to `v2.0.0` what's the plan for doing this?

Reply:
I have ideas on how to do this with python's `mro`